crates/sui-core/src/checkpoints/checkpoint_executor/data_ingestion_handler.rs

```diff
@@ -0,0 +1,152 @@
+            .collect::<HashSet<_>>();
+
+        let input_object_keys = fx
+            .input_shared_objects()
+            .into_iter()
+            .map(|kind| {
+                let (id, version) = kind.id_and_version();
+                ObjectKey(id, version)
+            })
+            .chain(
+                fx.modified_at_versions()
+                    .into_iter()
+                    .map(|(object_id, version)| ObjectKey(object_id, version)),
+            )
+            .collect::<HashSet<_>>()
+            .into_iter()
+            // Unwrapped-then-deleted objects are not stored in state before the tx, so we have nothing to fetch.
+            .filter(|key| !unwrapped_then_deleted_obj_ids.contains(&key.0))
+            .collect::<Vec<_>>();
+
+        let input_objects = authority_store
+            .multi_get_object_by_key(&input_object_keys)?
+            .into_iter()
+            .zip(&input_object_keys)
+            .map(|(object, object_key)| {
+                object.ok_or(SuiError::UserInputError {
+                    error: UserInputError::ObjectNotFound {
+                        object_id: object_key.0,
+                        version: Some(object_key.1),
+                    },
+                })
+            })
+            .collect::<SuiResult<Vec<_>>>()?;
+
+        let output_object_keys = fx
+            .all_changed_objects()
+            .into_iter()
+            .map(|(object_ref, _owner, _kind)| ObjectKey::from(object_ref))
+            .collect::<Vec<_>>();
+
+        let output_objects = authority_store
+            .multi_get_object_by_key(&output_object_keys)?
+            .into_iter()
+            .zip(&output_object_keys)
+            .map(|(object, object_key)| {
+                object.ok_or(SuiError::UserInputError {
+                    error: UserInputError::ObjectNotFound {
+                        object_id: object_key.0,
+                        version: Some(object_key.1),
+                    },
+                })
+            })
+            .collect::<SuiResult<Vec<_>>>()?;
+
+        let full_transaction = CheckpointTransaction {
+            transaction: tx.into(),
+            effects: fx,
+            events,
+            input_objects,
+            output_objects,
+        };
+        full_transactions.push(full_transaction);
+    }
+    let file_name = format!("{}.chk", checkpoint.sequence_number);
+    let checkpoint_data = CheckpointData {
+        checkpoint_summary: checkpoint.into(),
+        checkpoint_contents,
+        transactions: full_transactions,
+    };
+
+    std::fs::create_dir_all(&path).map_err(|err| {
+        SuiError::FileIOError(format!(
+            "failed to save full checkpoint content locally {:?}",
+            err
+        ))
+    })?;
+
+    bcs::serialize_into(
+        &mut std::fs::File::create(path.join(file_name)).map_err(|_| {
+            SuiError::FileIOError("failed to save full checkpoint content locally".to_string())
```

Comment:
Could we get into a state where downstream doesn't consume these files on time and we overload the filesystem to cause perf degradation. It may not happen but do you think it might make sense to put some upper limit on max pending files and refusing to create more

Reply:
Yeah we probably want to think about back-pressure here and if it also causes the fullnode to stop executing more checkpoints till things are cleared up.

Reply:
we create around 86400 files per day, which isn't much. Realistically, the indexer daemon would need to be completely down for a while, and that should trigger an alarm. 
We can pause processing checkpoints on the indexer FN until the backlog clears or even just panic, as it indicates a problem with the indexer daemon anyway

Reply:
@phoenix-o if you think backpressure makes sense, a simple way to do (this is just a suggestion) might be to set up directory notification for delete events and keep a counter of total deletes to know num files since start

Reply:
interesting, will take a look. Why not just count number of files directly(or probabilistically every other checkpoint)?

Reply:
Yeah that definitely works and an equally good solution. I was just trying to avoid polling at some time period but it should not be a problem if we do that too.

Reply:
Let's move the backpressure implementation into a separate PR. It seems like it deserves its own discussion